### PR TITLE
Refactor `warn_if_nan`

### PR DESCRIPTION
**Pull request by alicanb**
`warn_if_nan` was multiply defined across the codebase. This PR refactors them in single function in `pyro.util`. One significant functional change is `torch_isinf` and `torch.isnan` returns true even if a single element in tensor is inf or nan.

### CODE REVIEW

File: pyro/util.py [hunk 1]
```diff
@@ -30,20 +30,34 @@ def set_rng_seed(rng_seed):
...
+    return (x == float('inf')).any()
+
+
+def warn_if_nan(name, value):
+    """
+    A convenient function to warn if a Tensor or its grad contains any nan or inf,
+    also works with numbers.
+    """
+    if torch.is_tensor(value) and value.requires_grad:
+            value.register_hook(lambda x: warn_if_nan(name, x))
```
Comment (neerajprad) [inline pyro/util.py]: Nice to have this warning on NaN gradients. Should have 4 spaces here; otherwise this will fail the lint check.

File: pyro/util.py [hunk 2]
```diff
@@ -30,20 +30,34 @@ def set_rng_seed(rng_seed):
...
-    A convenient function to check if a Tensor contains all inf; also works with numbers
+    A convenient function to check if a Tensor contains any inf; also works with numbers
     """
     if isinstance(x, numbers.Number):
         return x == float('inf')
-    return (x == float('inf')).all()
+    return (x == float('inf')).any()
+
+
+def warn_if_nan(name, value):
```
Comment (neerajprad) [inline pyro/util.py]: If we have use cases where it is fine to have `+inf` values but not `NaN`s, then it would make sense to have an optional kwarg `warn_if_posinf=True`, so that optionally turn off the check on `+inf` values. That said, if you haven't found any code / failing tests yet, we can address this later.

Comment (neerajprad) [reply pyro/util.py]: What do you think about changing `name` to an optional but more general `msg=""` kwarg, which could be specified as `msg="log_prob_sum at site {}".format(name)`, or `msg="site: {}".format(name)`, and used as follows below:

```
warnings.warn("Encountered inf. " + msg)
```

File: pyro/util.py [hunk 3]
```diff
@@ -30,20 +30,34 @@ def set_rng_seed(rng_seed):
...
+    """
+    A convenient function to warn if a Tensor or its grad contains any nan or inf,
+    also works with numbers.
+    """
+    if torch.is_tensor(value) and value.requires_grad:
+            value.register_hook(lambda x: warn_if_nan(name, x))
+    if torch_isnan(value):
+        warnings.warn("Encountered NAN in {}".format(name))
+    if torch_isinf(value) and value > 0:
```
Comment (fritzo) [inline pyro/util.py]: `warn_if_nan()` should not warn if inf. Instead, create a separate `warn_if_inf()` method. In particular, most users of `warn_if_nan()` will want to allow `-inf` and many will want to allow `inf`.

File: pyro/util.py [hunk 4]
```diff
@@ -30,20 +30,34 @@ def set_rng_seed(rng_seed):
...
+
+def warn_if_nan(name, value):
+    """
+    A convenient function to warn if a Tensor or its grad contains any nan or inf,
+    also works with numbers.
+    """
+    if torch.is_tensor(value) and value.requires_grad:
+            value.register_hook(lambda x: warn_if_nan(name, x))
+    if torch_isnan(value):
+        warnings.warn("Encountered NAN in {}".format(name))
```
Comment (jpchen) [inline pyro/util.py]: is this actually being tested? where is `warnings` being imported?

**Comment by alicanb** [comment]
Hey guys, I addressed most of the reviews:
- new method: `warn_if_inf(value, msg, allow_neg_inf,allow_posinf)`
- `torch_isinf` checks for both +inf and -inf
- added `test_util.py` to test suite (only checks `warn_if_inf` and `warn_if_nan` for now)

File: pyro/util.py [hunk 5]
```diff
@@ -30,20 +30,46 @@ def set_rng_seed(rng_seed):
...
+    if torch_isnan(value):
+        warnings.warn("Encountered NaN{}".format((': ' if msg else '.') + msg))
+
+
+def warn_if_inf(value, msg="", allow_posinf=False, allow_neginf=False):
+    """
+    A convenient function to warn if a Tensor or its grad contains any inf,
+    also works with numbers.
+    """
+    if torch.is_tensor(value) and value.requires_grad:
```
Comment (jpchen) [inline pyro/util.py]: nit: to prevent needless memory allocation you can `return` or throw an error if both `allow_posinf` and `allow_neginf` are true

**Comment by fritzo** [comment]
Thanks for cleaning this up, @alicanb!
